### ISSUE TITLE
Second property change on context errors because applyPrevious is `null`

### ISSUE THREAD

**Issue by BigAB** [sindresorhus/on-change#34]
So this simple example of an object mutating itself, fails with the Error:
`TypeError: Cannot set property 'count' of null`

```
let count = 0;

const counterObj = {
  onTick() {
    this.count = count++;
  },
};

var counter = onChange(counterObj, (property, value) => {
  console.log(property, value);
});

counter.onTick();
counter.onTick();

```

I think the problem is in here https://github.com/sindresorhus/on-change/blob/master/index.js#L188-L212

...but I am having trouble following what is going on there.

It seems like `applyPrevious` is not handled correctly, and is resulting in `null` being set it `item` on https://github.com/sindresorhus/on-change/blob/master/index.js#L65

Which results in the error.

Just saw this https://github.com/sindresorhus/on-change/issues/32
Probably related, but just in case I will open this issue

**Comment by mjstahl** [comment]
I believe the issue lies in https://github.com/sindresorhus/on-change/blob/master/index.js#L191-L201

When in an `apply` trap, `applyPrevious` is not getting set.  I just can't figure out what it should get set to (i.e. I don't understand the purpose of `applyPrevious` in the codebase).

Personally I don't understand why it is set for Arrays, and for Dates, but not for normal objects.  I guess this goes back to not understanding its purpose. I will have to look investigate further.

**Comment by DarrenPaulWright** [comment]
Fixed by 2c8d1e0e4fceca3c4c3ad0add0b01650d2f2c6fa
